### PR TITLE
Add PR template checklist and guidelines

### PR DESCRIPTION
**Pull request by JenkoB**
With an aim to make writing and reviewing PRs easier and requirements more clear I have made PR checklist and guidelines for Resolwe-bio repository.

Please comment and contribute.
Thanks

### CODE REVIEW

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+## Checklist
+<!--
+Mark an `[x]` for completed items.
+-->
+
+* [ ] Update CHANGELOG.rst  for each commit separately.
+* [ ] Bump the process version:
+    * **Last number**: changes that do not affect the api/interface. Examples:
+        bugfix, typo fix, change/add warning messages,..
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: Should be three dots `...` not `,..`

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    * **Middle number**: bigger changes that don't make it backwards incompatible.
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: don't make process backwards incompatible.

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        Examples: add output field, add non-mandatory input parameter, use
+        a different tool that produces same results...
+    * **First number**: Backwards incompatible changes, are changes that
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: Backwards incompatible changes, (changes that break the api/interface).

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+* Code that was once accepted should be kept like it is. It should be changed
+    only when it's functionality is changed, or small parts may be changed
+    to fix bugs (If you change it only to make it nicer there is a chance
+    to introduce a bug).
+* Set the optimal number of cores that the tool can use (check if parallelization
+    can be performed).
+* Description of tool should be self explanatory with links to tool homepage
+    and publication.
+* Write explanatory input labels and add short commands in [ ] (for example:
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: Write explanatory input labels and add command line parameter names in square brackets (for example: Maximum length [`--maxlength`]).

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
+## Checklist
+<!--
+Mark an `[x]` for completed items.
+-->
+
+* [ ] Update CHANGELOG.rst  for each commit separately.
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: - Pay attention to always write entries under the "Unreleased" section
- Mark all breaking changes as `**BACKWARD INCOMPATIBLE:**` and put them before non-breaking changes

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+### Commit and PR messages ###
+
+* Each commit should be minimal (i.e. 1 change) and self-contained (including
+    tests).
+* Changes in docker images needs to be in separate PR.
+* For work in progress add [WIP] at the beginning of commit messages and
+    PR title.
+
+### Writing tests ###
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: IMO, "Tests" is sufficient heading.

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+* Add meaningful names of test files with spaces:
+    * Genomes: ``genome species.fa.gz``
+    * Annotations: ``annotation species.gtf.gz``
+    * Adapters: ``adapters-name/type-source``.
+* Output and specific input files should have name of the tested tool.
+* Keep small sizes of test files.
+* When testing workflows, verify workflow success as best as possible.
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: Add something like: "Pay attention to edge cases: What will happen is any data is missing, if length of some list is 0, if string is given where one expects integer, if file is empty, ... "

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+* Add ``re-progress`` notes.
+
+### Commit and PR messages ###
+
+* Each commit should be minimal (i.e. 1 change) and self-contained (including
+    tests).
+* Changes in docker images needs to be in separate PR.
+* For work in progress add [WIP] at the beginning of commit messages and
+    PR title.
+
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: Do not paste links to private repositories from the public one.

**Comment by JenkoB** [comment]
Thanks to Jure, Janez and Tadej for contribution. I have corrected text, by implementing all comments. Please make another round of reviewing.

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+* [ ] Bump the process version:
+  * **MAJOR version (first number)**: Backwards incompatible changes (changes
+    that break the api/interface). Examples: renaming the input/output, adding
+    mandatory input, removing input/output…
+  * **MINOR version (middle number)**: add functionality or changes in a
+    backwards-compatible manner. Examples: add output field, add non-mandatory
+    input parameter, use a different tool that produces same results...
+  * **PATCH version (last number)**: changes/bug fixes that do not affect
+    the api/interface. Examples: typo fix, change/add warning messages...
+* [ ] All inputs are used in process.
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: This line and the one below should go under processes section.

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+* [ ] All output fields have their ``re-save``.
+
+## Additional guidelines
+
+### Processes ###
+
+* Code that was once accepted should be kept like it is. It should be changed
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: This is general gudeline, appicable also to code outside processes.

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+-->
+
+* [ ] Update CHANGELOG.rst for each commit separately:
+  * Pay attention to write entries under the "Unreleased" section.
+  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
+    before non-breaking changes.
+  * If a commit modifies a feature listed under "Unreleased" section,
+    it might be sufficient to modify the existing CHANGELOG entry from previous
+    commit(s). When making a new CHANGELOG entry, augment it.
+* [ ] Bump the process version:
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: This should go under processes section

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
+## Checklist
+<!--
+Mark an `[x]` for completed items.
+-->
+
+* [ ] Update CHANGELOG.rst for each commit separately:
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: This could go under Commits and PR section

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  * Genomes: ``genome species.fa.gz``
+  * Annotations: ``annotation species.gtf.gz``
+  * Adapters: ``adapters-name/type-source``.
+* Output and specific input files should include the name of the tested tool.
+* Keep test files' size small. If there is no other option, add the large
+  test file to the ``files/large`` directory so it will be tracked with Git LFS.
+* Pay attention to edge cases: what will happen if any data is
+  missing, if length of some list is 0, if string is given where one expects
+  integer, if file is empty, ...
+* When testing workflows, verify workflow success as best as possible.
```
Comment (JureZmrzlikar) [inline docs/PULL_REQUEST_TEMPLATE.md]: Maybe rephrase: "When testing workflows, try to verify workflow success by checking one of the last steps. Pick the one that is most likely to fail. Also test that all steps have `'OK'` status.

File: docs/PULL_REQUEST_TEMPLATE.md (hunk 12 above)
Comment (tjanez) [reply docs/PULL_REQUEST_TEMPLATE.md]: @JureZmrzlikar , these items are meant to be a check list, therefore they are all the beginning under the Checklist section.

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+Mark an `[x]` for completed items.
+-->
+
+* [ ] Update CHANGELOG.rst for each commit separately:
+  * Pay attention to write entries under the "Unreleased" section.
+  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
+    before non-breaking changes.
+  * If a commit modifies a feature listed under "Unreleased" section,
+    it might be sufficient to modify the existing CHANGELOG entry from previous
+    commit(s). When making a new CHANGELOG entry, augment it.
```
Comment (tjanez) [inline docs/PULL_REQUEST_TEMPLATE.md]: Remove "When making a new CHANGELOG entry, augment it." sentence. It is obsolete now.

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+* [ ] Bump the process version:
+  * **MAJOR version (first number)**: Backwards incompatible changes (changes
```
Comment (tjanez) [inline docs/PULL_REQUEST_TEMPLATE.md]: Backwards -> Backward

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    that break the api/interface). Examples: renaming the input/output, adding
```
Comment (tjanez) [inline docs/PULL_REQUEST_TEMPLATE.md]: api/interface -> API
renaming the input/output -> renaming input/output

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    mandatory input, removing input/output…
+  * **MINOR version (middle number)**: add functionality or changes in a
+    backwards-compatible manner. Examples: add output field, add non-mandatory
+    input parameter, use a different tool that produces same results...
+  * **PATCH version (last number)**: changes/bug fixes that do not affect
+    the api/interface. Examples: typo fix, change/add warning messages...
+* [ ] All inputs are used in process.
+* [ ] All output fields have their ``re-save``.
```
Comment (tjanez) [inline docs/PULL_REQUEST_TEMPLATE.md]: have their ``re-save`` -> are referenced in ``re-save`` calls

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+* Set the optimal number of cores that the tool can use (check if parallelization
+  can be performed).
+* Description of tool should be self explanatory with links to tool's homepage
+  and publication.
+* Write explanatory input labels and add short command line parameter names
+  in square brackets (for example: Maximum length [``--maxlength``]).
+* Where possible provide default values for the process input fields.
+  If they are provided, then ``required: false`` is not needed.
+* Add descriptive error and warning messages.
+* Place a ``re-checkrc`` call after each process step (call of a tool, Bash
```
Comment (tjanez) [inline docs/PULL_REQUEST_TEMPLATE.md]: Bash -> a Bash

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+  command a Python script or an R script).
```
Comment (tjanez) [inline docs/PULL_REQUEST_TEMPLATE.md]: command -> command,

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+    that break the api/interface). Examples: renaming the input/output, adding
+    mandatory input, removing input/output…
+  * **MINOR version (middle number)**: add functionality or changes in a
+    backwards-compatible manner. Examples: add output field, add non-mandatory
+    input parameter, use a different tool that produces same results...
+  * **PATCH version (last number)**: changes/bug fixes that do not affect
+    the api/interface. Examples: typo fix, change/add warning messages...
+* [ ] All inputs are used in process.
+* [ ] All output fields have their ``re-save``.
+
```
Comment (tjanez) [inline docs/PULL_REQUEST_TEMPLATE.md]: Add a comment indicating that the following lines should be read and discarded before opening a PR.

Something like:
```
<!--
Read the following guidelines and remove them before opening the pull request.
-->
```

File: docs/PULL_REQUEST_TEMPLATE.md (hunk 12 above)
Comment (JenkoB) [reply docs/PULL_REQUEST_TEMPLATE.md]: For now I will leave only these 4 points as Checklist and the others as Guidelines. If we will notice this is not the optimal solution we can change it.

File: docs/PULL_REQUEST_TEMPLATE.md [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+  * Pay attention to write entries under the "Unreleased" section.
+  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
+    before non-breaking changes.
+  * If a commit modifies a feature listed under "Unreleased" section,
+    it might be sufficient to modify the existing CHANGELOG entry from previous
+    commit(s).
+* [ ] Bump the process version:
+  * **MAJOR version (first number)**: Backward incompatible changes (changes
+    that break the api/interface). Examples: renaming the input/output, adding
+    mandatory input, removing input/output…
```
Comment (tjanez) [inline docs/PULL_REQUEST_TEMPLATE.md]: Replace `…` with `...`.
